src/main.ts

```diff
@@ -0,0 +1,794 @@
+    ) {
+      throw new Error(
+        "Caching of unstable releases is not supported on macOS runners"
+      );
+    }
+
+    // Validate authentication
+    validateAuth(config);
+
+    // Resolve version
+    config.resolvedVersion = await resolveVersion(config.version, runnerOS);
+    core.info(`Resolved Tailscale version: ${config.resolvedVersion}`);
+
+    // Set architecture
+    config.arch = getTailscaleArch(runnerOS);
+
+    // Install Tailscale
+    await installTailscale(config, runnerOS);
+
+    // Start daemon (non-Windows only)
+    if (runnerOS !== runnerWindows) {
+      await startTailscaleDaemon(config);
+    }
+
+    // Connect to Tailscale
+    await connectToTailscale(config, runnerOS);
+
+    // Check Tailscale status (cross-platform)
+    try {
+      const status = await getTailscaleStatus();
+      core.debug(`Tailscale status: ${JSON.stringify(status)}`);
+      if (status.BackendState === "Running") {
+        core.info("✅ Tailscale is running and connected!");
+        await pingHostsIfNecessary(config);
+        // Explicitly exit to prevent hanging
+        process.exit(0);
+      } else {
+        core.setFailed(`❌ Tailscale backend state: ${status.BackendState}`);
+        process.exit(1);
+      }
+    } catch (err) {
+      core.warning(`Failed to get Tailscale status: ${err}`);
+      // Still exit successfully since the main connection worked
+      core.info("✅ Tailscale daemon is connected!");
+      await pingHostsIfNecessary(config);
+      // Explicitly exit to prevent hanging
+      process.exit(0);
+    }
+  } catch (error) {
+    core.setFailed(error instanceof Error ? error.message : String(error));
+  }
+}
+
+async function pingHostsIfNecessary(config: TailscaleConfig): Promise<void> {
+  if (config.pingHosts.length == 0) {
+    return;
+  }
+
+  core.info(
+    `Will ping hosts ${config.pingHosts.join(
+      ","
+    )} up to 3 minutes each in order to check connectivity`
+  );
+  for (const host of config.pingHosts) {
+    await pingHost(host);
+  }
+}
+
+async function pingHost(host: string): Promise<void> {
+  core.info(`Pinging host ${host}`);
+  for (let i = 0; i <= 36; i++) {
+    if (i < 0) {
+      // wait 5 seconds before pinging
+      wait(5000);
+    }
+    let result = await exec.getExecOutput(
+      cmdTailscale,
+      ["ping", "-c", "1", host],
+      { ignoreReturnCode: true }
+    );
+    if (result.exitCode === 0) {
+      core.info(`✅ Ping host ${host} reachable via direct connection!`);
+      return;
+    } else if (result.stderr.includes("direct connection not established")) {
+      core.info(`✅ Ping host ${host} reachable via DERP!`);
+      return;
+    }
+  }
+  core.setFailed(`❌ Ping host ${host} did not respond`);
+  process.exit(1);
+}
+
+async function getInputs(): Promise<TailscaleConfig> {
+  let ping = core.getInput("ping");
+  let pingHosts = ping?.length > 0 ? ping.split(",") : [];
+
+  return {
+    version: core.getInput("version") || "1.82.0",
```

Comment:
Think we'll want to replace `1.82.0` with `1.88.3` throughout now that we've bumped this version in the v3 line.